### ISSUE TITLE
Inconsistency in Type of Tensor when using dataset.take() vs dataset.map()

### ISSUE THREAD

**Issue by shun-lin** [tensorflow/datasets#1298]
**Short description**
I do not know if this is a bug or a feature, but I open this issue in case others are experiencing the same thing.

When I query the tensors within the datasets loaded using `tfds.load` using `.take()`, they are `EagerTensor`, which allows me to query the value like `tensor.numpy()`. But when I try to apply a function to the entire dataset using `dataset.map()`, the tensors are of type `Tensor`, and therefore I cannot use `.numpy()` to retrieve the value and instead I need to start a `session` and then use `eval()`, which is not ideal.

My use case is in the Additional Context below.

**Environment information**
* Operating System: gLinux, but reproducible on Colab
* Python version: 3.6.9
* `tensorflow-datasets`/`tfds-nightly` version: 1.3.2
* `tensorflow`/`tensorflow-gpu`/`tf-nightly`/`tf-nightly-gpu` version: 2.1.0-rc1
**Reproduction instructions**

I have made a Colab with a minimal code that reproduce the inconsistency (using tfds mnist dataset):
https://colab.research.google.com/drive/1LjezThWkxazBmfayPZyWgWwmmZffWUnP?authuser=1#scrollTo=Ft6wp2JrMn6Q

**Link to logs**
See colab linked above.

**Expected behavior**
Since tf 2.x Eager Execution is enabled by default, I am expecting that when I use dataset.map(), the tensor insides are EagerTensor as well. But there may be reasons why this should not be the case, if there are any such reasons please let me know!

**Additional context**
For my use case I want to randomly select a few classes of Labeled Faces in the Wild for testing and training. Because the labels in the dataset are `String`, I need to first map them to `integer`, random select some of those integers, and then apply `filter` to get the testing and training dataset. To build the mapping from the `String` labels to `int`, I need the value within the `label` Tensor as `Tensor` is not hashable.

EDIT: adding more additional context.

Ideally I do not want to read the lfw dataset into memory as it's pretty big (20GB).

Thank you so much for your time!

**Comment by Conchylicultor** [comment]
The colab is not public, so I can't really see.

The reason is that `tf.data.Dataset` pipelines are designed for performance. All the input pipeline is compiled in graph and run in parallel. So the map function cannot run arbitrary python code.

For insert Python code inside a TF graph, you could use `tf.py_function`, which will convert tensors to np and vice verse. See https://www.tensorflow.org/api_docs/python/tf/py_function
But because of the conversion tf.Tensor -> np -> tf.Tensor, be aware that it may significantly affect the input performances.

**Comment by shun-lin** [comment]
thank you so much for your quick response @Conchylicultor , I will try `tf.py_function` and `tf.lookup` :)
